spark/src/main/scala/geotrellis/spark/io/LayerUpdater.scala

```diff
@@ -2,18 +2,57 @@ package geotrellis.spark.io
 
 import geotrellis.spark._
 
-import geotrellis.spark.Boundable
+import geotrellis.spark._
+import geotrellis.spark.io.avro._
+import geotrellis.spark.io.avro.codecs._
+import geotrellis.spark.io.json._
+
+import org.apache.avro.Schema
 import org.apache.spark.rdd.RDD
+import spray.json._
 
-abstract class LayerUpdater[ID, K: Boundable, V, M] {
-  type Container = RDD[(K, V)] with Metadata[M]
+import scala.reflect.ClassTag
 
-  def update(id: ID, rdd: Container): Unit
+abstract class LayerUpdater[ID] {
+  protected def _update[
+    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
+    V: AvroRecordCodec: ClassTag,
+    M: JsonFormat: Component[?, Bounds[K]]
+  ](id: ID, rdd: RDD[(K, V)] with Metadata[M], keyBounds: KeyBounds[K]): Unit
 
-  def mergeUpdate(id: ID, reader: FilteringLayerReader[ID, K, M, Container], rdd: Container)
-                 (merge: (Container, Container) => Container) = {
-    val bounds = implicitly[Boundable[K]].collectBounds(rdd)
-    val existing = reader.query(id).where(Intersects(bounds)).toRDD
-    update(id, merge(existing, rdd))
+  protected def schemaHasChanged[K: AvroRecordCodec, V: AvroRecordCodec](writerSchema: Schema): Boolean = {
```

Comment:
`fingerprintMatches` was based on [this documentation](https://avro.apache.org/docs/1.7.6/spec.html#Schema+Fingerprints), and I think is more generally useful than this case (although it is currently only used in this case). `schemaHasChanged` deals explicitly with the way writers pull together the K and V codecs to get the schema, so I think this separation is warranted.